src/qiboml/optimizers/heuristics.py

```diff
@@ -0,0 +1,195 @@
+
+    Args:
+        verbosity (Optional[bool]): verbosity level of the optimization. If `True`, logging messages are displayed.
+        sigma0 (Optional[float]): scalar, initial standard deviation in each coordinate.
+            sigma0 should be about 1/4th of the search domain width (where the
+            optimum is to be expected).
+        restarts (Optional[int]): number of restarts with increasing population size.
+        restart_from_best (Optional[bool]): which point to restart from.
+        iconpopsize (Optional[int]): multiplier for increasing the population size popsize before each restart.
+        callback (Optional[callable]): a callable called after each optimization iteration.
+    """
+
+    sigma0: Optional[float] = field(default=0.5)
+    restarts: Optional[int] = field(default=0)
+    restarts_from_best: Optional[bool] = field(default=False)
+    iconpopsize: Optional[int] = field(default=2)
+    callback: Optional[callable] = field(default=None)
+
+    def __str__(self):
+        return "cmaes"
+
+    def show_fit_options(self, keyword: str = None):
+        """
+        Return all the available fit options for the optimizer.
+
+        Args:
+            keyword (str): keyword to help the research of fit options into the
+                options dictionary.
+        """
+        return cma.CMAOptions(keyword)
+
+    def fit(
+        self,
+        initial_parameters: Union[List, ndarray],
+        loss: callable,
+        args: Optional[Tuple] = None,
+        fit_options: Optional[dict] = None,
+    ):
+        """Perform the optimizations via CMA-ES.
+
+        Args:
+            initial_parameters (np.ndarray or list): array with initial values
+                for gate parameters.
+            loss (callable): loss function to train on.
+            args (tuple): tuple containing loss function arguments.
+            fit_options (dict): fit extra options. To have a look to all
+                possible options please use `CMAES.show_fit_options()`.
+
+        Returns:
+            tuple: best loss value (float), best parameter values (np.ndarray), full cma result object.
+        """
+
+        if fit_options is None:
+            options = {}
+        else:
+            options = fit_options
+
+        log.info(f"Optimization is performed using the optimizer: {self.__str__()}")
+
+        r = cma.fmin2(
+            objective_function=loss,
+            x0=initial_parameters,
+            args=args,
+            sigma0=self.sigma0,
+            restarts=self.restarts,
+            restart_from_best=self.restarts_from_best,
+            incpopsize=self.iconpopsize,
+            callback=self.callback,
+            options=options,
+        )
+
+        return r[1].result.fbest, r[1].result.xbest, r
+
+
+@dataclass
+class BasinHopping(Optimizer):
+    """
+    Global optimizer based on: https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.basinhopping.html.
```

Comment:
Other minor remark: for this kind of references, you could use intersphinx. I.e. just use the appropriate cross-reference:
```suggestion
    Global optimizer based on: :func:`scipy.optimize.basinhopping`.
```
and then we will configure Sphinx to link SciPy docs as well.